OsmAnd-java/src/main/java/net/osmand/router/RoutingContext.java

```diff
@@ -513,7 +539,103 @@ private long getRoutingTile(int x31, int y31, long memoryLimit) {
 	}
 
 	
+	private void connectPoint(final RoutingSubregionTile ts, RouteDataObject ro, List<DirectionPoint> points) {
+		for (DirectionPoint np : points) {
+			if (np.types.size() == 0) {
+				continue;
+			}	
+			
+			int wptX = MapUtils.get31TileNumberX(np.getLongitude());
+			int wptY = MapUtils.get31TileNumberY(np.getLatitude());
+			
+			int x = ro.getPoint31XTile(0);
+			int y = ro.getPoint31YTile(0);
+			
+			double mindist = config.directionPointsRadius * 2;
+			int indexToInsert = 0;
+			int mprojx = 0;
+			int mprojy = 0;
+			for (int i = 1; i < ro.getPointsLength(); i++) {
+				int nx = ro.getPoint31XTile(i);
+				int ny = ro.getPoint31YTile(i);
+				boolean sgnx = nx - wptX > 0;
+				boolean sgx = x - wptX > 0;
+				boolean sgny = ny - wptY > 0;
+				boolean sgy = y - wptY > 0;
+				boolean checkPreciseProjection = true;
+				if (sgny == sgy && sgx == sgnx) {
+					// Speedup: point outside of rect (line is diagonal) distance is likely be bigger
+					double dist = MapUtils.squareRootDist31(wptX, wptY, Math.abs(nx - wptX) < Math.abs(x - wptX) ? nx : x,
+							Math.abs(ny - wptY) < Math.abs(y - wptY) ? ny : y);
+					checkPreciseProjection = dist < config.directionPointsRadius;
+				} 
+				if (checkPreciseProjection) {
+					QuadPoint pnt = MapUtils.getProjectionPoint31(wptX, wptY, x, y, nx, ny);
+					int projx = (int) pnt.x;
+					int projy = (int) pnt.y;
+					double dist = MapUtils.squareRootDist31(wptX, wptY, projx, projy);
+					if (dist < mindist) {
+						indexToInsert = i;
+						mindist = dist;
+						mprojx = projx;
+						mprojy = projy;
+					}
+				}
+
+				
+				x = nx;
+				y = ny;
+			}
+			boolean sameRoadId = np.connected != null && np.connected.getId() == ro.getId();
+			boolean pointShouldBeAttachedByDist = (mindist < config.directionPointsRadius && mindist < np.distance);
+			
+			if (pointShouldBeAttachedByDist && !sameRoadId) {
+				//System.out.println(String.format("INSERT %s (%d-%d) %.0f m [%.5f, %.5f] - %d, %d ", ro, indexToInsert, indexToInsert + 1, mindist,
+				//		MapUtils.get31LongitudeX(wptX), MapUtils.get31LatitudeY(wptY), wptX, wptY));
+				if (np.connected != null) {
+					// clear old connected points
+					int pointIndex = -1;
+					for (int i = 0; i < np.connected.getPointsLength(); i++) {
+						int tx = np.connected.getPoint31XTile(i);
```

Comment:
Used comment to check if the road has point with same coordinates, now it's gone and it doesn't look like the code works correctly if 
mprojx, mprojy was present already on the road. 
So it looks buggy deleting / inserting duplicate points.

Reply:
We using the checking !sameRoadId, so it will not insert twice

Reply:
not related to sameRoadId: if the road has point with same coordinates, now it's gone and it doesn't look like the code works correctly if mprojx, mprojy was present already on the road (! point was present without any modification!).